Lombiq.Tests.UI.Samples/Tests/BasicVisualVerificationTests.cs

```diff
@@ -0,0 +1,50 @@
+using Lombiq.Tests.UI.Attributes;
+using Lombiq.Tests.UI.Constants;
+using Lombiq.Tests.UI.Extensions;
+using Lombiq.Tests.UI.Services;
+using OpenQA.Selenium;
+using System.Drawing;
+using System.Threading.Tasks;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Lombiq.Tests.UI.Samples.Tests;
+
+// In this basic test we will check visually the rendered content.
+public class BasicVisualVerificationTests : UITestBase
+{
+    public BasicVisualVerificationTests(ITestOutputHelper testOutputHelper)
+        : base(testOutputHelper)
+    {
+    }
+
+    // Checking that everything is OK with the branding of the navbar on the homepage.
+    // For this magic we are using the ImageSharp.Compare package. You can find more info about it here:
+    // https://github.com/Codeuctivity/ImageSharp.Compare/blob/2.0.46/README.md.
+    [Theory, Chrome, VisualVerification]
+    public Task VerifyNavbar(Browser browser) =>
+        ExecuteTestAfterSetupAsync(
+            context =>
+            {
+                // Instead of using context.SetBrowserSize(...), witch is sets the outer size of the browsers window to
+                // the given size, we are using context.SetViewportSize(...) here. This is because the windows borders,
+                // toolbars, tabs and scroll bars usually have different sizes on different platforms/browsers, but we
+                // want the same geometries of rendered content on all platforms.
+                context.SetViewportSize(CommonDisplayResolutions.HdPlus);
+
+                var navbarElementSelector = By.ClassName("navbar-brand");
+                // We set the browser's window size, DPI, and scale settings explicitly to make the test environment
+                // similar on every platform, but we have no control over the fonts. In this case, the selected block
+                // element size, mainly height, is depends on the font used on a given platform.
+                // We don't know the exact size of the selected element, so we should select the interesting region for
+                // reference and be captured as bitmap, and use it at the end. Without cropping the reference and
+                // captured bitmap to the same size, the ImageSharpCompare.CalcDiff(...) throws an
+                // ImageSharpCompareException.
+                var cropRegion = new Rectangle(0, 0, 1400, 23);
+
+                context.VisualAssertApproved(navbarElementSelector, 8, cropRegion);
+            },
```

Comment:
I get this if I run `context.VisualAssertApproved(By.CssSelector("body"), 0);` after this:

![image](https://user-images.githubusercontent.com/1976647/172931265-ffe8d78a-316e-4518-8e89-7cd4bf469aef.png)

Multiple verifications should be possible in the same test. Perhaps add a suffix to the verification file generated from the `By`? Or if that doesn't work, add an overload where you can specify a name, to distinguish between multiple verifications within the same test?

Reply:
There is a `configuration` callback where you can configure a prefix.
Example: https://github.com/Lombiq/Orchard-Chart.js/blob/624f4c9a455e7708750cd7777d4cf278f9bc788d/Lombiq.ChartJs.Tests.UI/Extensions/TestCaseUITestContextExtensions.cs#L54

Reply:
Right. However, the exception is not very helpful. And how about using the `By`?

Reply:
I'll check what can I do with the exception.

`By`:
 - the [`By.ToString()`](https://github.com/SeleniumHQ/selenium/blob/9a52775cc4142b148ed8d2d8922340411b35235d/dotnet/src/webdriver/By.cs#L356) returns something like[ `"By.XPath: " + xpathToFind`](https://github.com/SeleniumHQ/selenium/blob/9a52775cc4142b148ed8d2d8922340411b35235d/dotnet/src/webdriver/By.cs#L244) or [`"By.Id: " + idToFind`](https://github.com/SeleniumHQ/selenium/blob/9a52775cc4142b148ed8d2d8922340411b35235d/dotnet/src/webdriver/By.cs#L180), etc, so this is not usable as a filename directly. Creating custom formatter for `By` is a possible solution.
 - we have overloads with `IWebElement element` instead of `By elementSelector`, so we need `configuration.WithDumpFileNamePrefix(...)` anyway

Reply:
One more, a possible test case:
 - verify element content with selector `By.Id("container")`
 - click or any other user action
 - verify element content again with the same selector `By.Id("container")`
In this case, prefixing with `By` will not help us out.

Reply:
Yes, and I don't see a way around that apart from doing some complex reflection thing. However, having an approach that doesn't work 100% of the time shouldn't prevent us from doing something that works most of the time.

Having more than one assertion for specific, different elements in a single test, as well as having automatic assertions for all pages in a given set will be the most common use-cases if other assertions we do in existing tests are a usable analog. If this works easily then it's no problem that we have to do something special for the remainder of the cases.